File: src/python_adapter/model_discrepancy/MD_Data_Interface_Py.m
classdef MD_Data_Interface_Py < MD_Data_Interface

    properties
        data_interface_py
    end

    methods

        function [u_opt] = Load_Optimal_u(this)
            u_opt = this.data_interface_py.Load_Optimal_u();
            u_opt = double(u_opt);
            if size(u_opt, 1) == 1
                u_opt = u_opt';
            end
        end

        function [z_opt] = Load_Optimal_z(this)
            z_opt = this.data_interface_py.Load_Optimal_z();
            z_opt = double(z_opt);
            if size(z_opt, 1) == 1
                z_opt = z_opt';
            end
        end

        function [Z] = Load_Z_Data(this)
            Z = this.data_interface_py.Load_Z_Data();
            Z = double(Z);
            if size(Z, 1) == 1
                Z = Z';
            end
        end

        function [D] = Load_d_Data(this)
            D = this.data_interface_py.Load_d_Data();
            D = double(D);
            if size(D, 1) == 1
                D = D';
            end
        end

        function this = MD_Data_Interface_Py(data_interface_py)
            this.data_interface_py = data_interface_py;
        end

    end

end

File: src/python_adapter/model_discrepancy/interfaces/problem_interfaces/MD_Data_Interface_Py.py
import numpy as np
from abc import ABC, abstractmethod

class MD_Data_Interface_Py():

    def __init__(self):
        pass
    
    # -------------------------------------------------------------------------------------------------------
    # Abstract methods which must be implemented
    
    @abstractmethod
    def Load_Optimal_u_Py(self):
        pass

    @abstractmethod
    def Load_Optimal_z_Py(self):
        pass

    # -------------------------------------------------------------------------------------------------------
    # Optional methods which must be implemented to enable some analyes

    def Load_Z_Data_Py(self):
        Z = []
        print('Load_Z_Data_Py must be implemented for any analyses except for optimal experimental design')
        return Z

    def Load_d_Data_Py(self):
        D = []
        print('Load_d_Data_Py must be implemented for any analyses except for optimal experimental design')
        return D

    # -------------------------------------------------------------------------------------------------------\
    # Matlab-Python interoperabilty methods

    def Load_Optimal_u(self):
        return self.Load_Optimal_u_Py()

    def Load_Optimal_z(self):
        return self.Load_Optimal_z_Py()

    def Load_Z_Data(self):
        Z = self.Load_Z_Data_Py()
        return Z

    def Load_d_Data(self):
        D = self.Load_d_Data_Py()
        return D
